lib/srv/discovery/fetchers/kube_services.go

```diff
@@ -0,0 +1,335 @@
+				if err != nil {
+					f.Log.WithError(err).Warnf("Could not get app from a Kubernetes service %q, port %d", service.GetName(), port.Port)
+					return nil
+				}
+
+				appsMu.Lock()
+				apps = append(apps, newApp)
+				appsMu.Unlock()
+			}
+
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return apps.AsResources(), nil
+}
+
+func (f *KubeAppFetcher) ResourceType() string {
+	return types.KindApp
+}
+
+func (f *KubeAppFetcher) Cloud() string {
+	return ""
+}
+
+func (f *KubeAppFetcher) String() string {
+	return fmt.Sprintf("KubeAppFetcher(Namespaces=%v, Labels=%v)", f.Namespaces, f.FilterLabels)
+}
+
+// autoProtocolDetection tries to determine port's protocol. It uses heuristics and port HTTP pinging if needed, provided
+// by protocol checker. It is used when no explicit annotation for port's protocol was provided.
+func autoProtocolDetection(serviceFQDN string, port v1.ServicePort, pc ProtocolChecker) string {
+	if port.AppProtocol != nil {
+		switch p := strings.ToLower(*port.AppProtocol); p {
+		case protoHTTP, protoHTTPS:
+			return p
+		}
+	}
+
+	if port.Port == 443 || strings.EqualFold(port.Name, protoHTTPS) {
+		return protoHTTPS
+	}
+
+	if pc != nil {
+		result := pc.CheckProtocol(fmt.Sprintf("%s:%d", serviceFQDN, port.Port))
+		if result != protoTCP {
+			return result
+		}
+	}
+
+	if port.Port == 80 || port.Port == 8080 || strings.EqualFold(port.Name, protoHTTP) {
+		return protoHTTP
+	}
+
+	return protoTCP
+}
+
+func getServiceFQDN(s v1.Service) string {
+	host := s.GetName()
+	if s.Spec.Type == v1.ServiceTypeExternalName {
+		host = s.Spec.ExternalName
+	}
+	return fmt.Sprintf("%s.%s.svc.cluster.local", host, s.GetNamespace())
+}
+
+// ProtocolChecker is an interface used to check what protocol uri serves
+type ProtocolChecker interface {
+	CheckProtocol(uri string) string
+}
+
+func getServicePorts(s v1.Service) ([]v1.ServicePort, error) {
+	preferredPort := ""
+	for k, v := range s.GetAnnotations() {
+		if k == types.DiscoveryPortLabel {
+			preferredPort = v
+		}
+	}
+	var availablePorts []v1.ServicePort
+	for _, p := range s.Spec.Ports {
+		// Only supporting TCP ports.
+		if p.Protocol != v1.ProtocolTCP {
+			continue
+		}
+		availablePorts = append(availablePorts, p)
+		// If preferred port is specified and we found it in available ports, use this one
+		if preferredPort != "" && (preferredPort == strconv.Itoa(int(p.Port)) || p.Name == preferredPort) {
+			return []v1.ServicePort{p}, nil
+		}
+	}
+
+	// If preferred port is specified and we're here, it means we couldn't find it in service's ports.
+	if preferredPort != "" {
+		return nil, trace.BadParameter("specified preferred port %s is absent among available service ports", preferredPort)
+	}
+
+	return availablePorts, nil
+}
+
+type ProtoChecker struct {
+	InsecureSkipVerify bool
+	client             *http.Client
+}
+
+func (p *ProtoChecker) CheckProtocol(uri string) string {
+	if p.client == nil {
```

Comment:
can we create the client when we instantiate the ProtoChecker?

Otherwise, this will trigger a race condition when multiple calls to `CheckProtocol` happen in //